state-surgery/migrate.go

```diff
@@ -0,0 +1,269 @@
+	outStateDB, err := state.New(common.Hash{}, state.NewDatabase(outDB), nil)
+	if err != nil {
+		log.Crit("error opening output state DB", "err", err)
+	}
+
+	// Dump the state and get the state root.
+	log.Info("dumping state")
+	newRoot := dumpState(inUnderlyingDB, inStateDB, outStateDB, stateRoot, genesis, expectedSupply)
+
+	// Now that the state is dumped, insert the genesis block. We pass in a nil
+	// database here because we don't want to update the state again with the
+	// pre-allocs.
+	//
+	// Unlike regular Geth (which panics if you try to import a genesis state with a nonzero
+	// block number), the block number can be anything.
+	block := genesis.ToBlock(nil)
+
+	// Geth block headers are immutable, so swap the root and make a new block with the
+	// updated root.
+	header := block.Header()
+	header.Root = newRoot
+	block = types.NewBlock(header, nil, nil, nil, trie.NewStackTrie(nil))
+	blob, err := json.Marshal(genesis)
+	if err != nil {
+		log.Crit("error marshaling genesis state", "err", err)
+	}
+
+	// Write the genesis state to the database. This is taken verbatim from Geth's
+	// core.Genesis struct.
+	rawdb.WriteGenesisState(outDB, block.Hash(), blob)
+	rawdb.WriteTd(outDB, block.Hash(), block.NumberU64(), block.Difficulty())
+	rawdb.WriteBlock(outDB, block)
+	rawdb.WriteReceipts(outDB, block.Hash(), block.NumberU64(), nil)
+	rawdb.WriteCanonicalHash(outDB, block.Hash(), block.NumberU64())
+	rawdb.WriteHeadBlockHash(outDB, block.Hash())
+	rawdb.WriteHeadFastBlockHash(outDB, block.Hash())
+	rawdb.WriteHeadHeaderHash(outDB, block.Hash())
+	rawdb.WriteChainConfig(outDB, block.Hash(), genesis.Config)
+	return nil
+}
+
+func dumpState(inDB state.Database, inStateDB *state.StateDB, outStateDB *state.StateDB, root common.Hash, genesis *core.Genesis, expectedSupply *big.Int) common.Hash {
+	// Open a trie based on the currently-configured state root.
+	tr, err := inDB.OpenTrie(root)
+	if err != nil {
+		panic(err)
+	}
+
+	// Iterate over the Genesis allocation accounts. These will override
+	// and accounts found in the state.
+	log.Info("importing allocated accounts")
+	for addr, account := range genesis.Alloc {
+		outStateDB.AddBalance(addr, account.Balance)
+		outStateDB.SetCode(addr, account.Code)
+		outStateDB.SetNonce(addr, account.Nonce)
+		for key, value := range account.Storage {
+			outStateDB.SetState(addr, key, value)
+		}
+		log.Info("allocated account", "addr", addr)
+	}
+
+	var (
+		accounts     uint64
+		lastAccounts uint64
+		start        = time.Now()
+		logged       = time.Now()
+	)
+	log.Info("Trie dumping started", "root", tr.Hash())
+
+	// Keep track of total OVM ETH migrated to verify that the script worked.
+	totalOVM := new(big.Int)
+
+	// Iterate over each account in the state.
+	it := trie.NewIterator(tr.NodeIterator(nil))
+	for it.Next() {
+		// It's up to use to decode trie data.
+		var data types.StateAccount
+		if err := rlp.DecodeBytes(it.Value, &data); err != nil {
+			panic(err)
+		}
```

Comment:
I believe some accounts may have a balance in the OVM eth contract, but are otherwise completely unused (zero nonce etc.) and don't show up in the account trie. So they would be missed during this iteration, and not get migrated over their balance, losing funds.

Reply:
Got it. The application does a check to make sure that the migrated balance matches the total supply of OVM ETH, so if any of these exist I should be able to catch them. If these addresses do exist I'll need to do some archaeology to dig up the actual addresses since we can't easily map between OVM ETH storage slots and addresses.

Reply:
I'm running this with the check active against mainnet right now; should have an answer in about an hour or so.